packages/compass-aggregations/src/components/aggregation-side-panel/stage-wizard-use-cases/match/group.tsx

```diff
@@ -0,0 +1,395 @@
+  dataTestId: string;
+  onAddNestedGroupClick: () => void;
+};
+
+export type GroupProps = {
+  fields: WizardComponentProps['fields'];
+  nestingLevel: number;
+  // See: GroupHeaderProps['disableAddGroup']
+  disableAddGroup: boolean;
+  // See: GroupHeaderProps['disableRemoveGroup']
+  disableRemoveGroup: boolean;
+  group: MatchConditionGroup;
+  onGroupChange: (changedGroup: MatchConditionGroup) => void;
+  onRemoveGroupClick: () => void;
+  onAddGroupClick: () => void;
+};
+
+export type CreateGroupFn = (
+  group?: Partial<MatchConditionGroup>
+) => MatchConditionGroup;
+
+// Helpers
+const MAX_ALLOWED_NESTING = 3;
+
+/**
+ * Returns a function to create a group with incremental ids. Consider using
+ * already created `createGroup` below instead of making another one yourself.
+ * This function is exported primarily to aid in testing.
+ */
+export const makeCreateGroup = (
+  createCondition: CreateConditionFn
+): CreateGroupFn => {
+  let id = 1;
+  return (
+    group: Omit<Partial<MatchConditionGroup>, 'id'> = {}
+  ): MatchConditionGroup => ({
+    id: id++,
+    logicalOperator: '$and',
+    conditions: [createCondition()],
+    groups: [],
+    ...group,
+  });
+};
+
+export const createGroup = makeCreateGroup(createCondition);
+
+export const makeGroupsHandlers = (
+  groups: MatchConditionGroups,
+  onGroupsChange: (newGroups: MatchConditionGroups) => void
+) => ({
+  handleGroupChange(groupIdx: number, newGroup: MatchConditionGroup) {
+    if (!groups[groupIdx]) {
+      return;
+    }
+
+    const newGroups = [...groups];
+    newGroups[groupIdx] = newGroup;
+    // Update the other group's logicalOperator
+    // if the current group operator is updated
+    const otherGroupIdx = groupIdx === 0 ? 1 : 0;
+    const otherGroup = newGroups[otherGroupIdx];
+    if (otherGroup && otherGroup.logicalOperator === newGroup.logicalOperator) {
+      otherGroup.logicalOperator =
+        newGroup.logicalOperator === '$and' ? '$or' : '$and';
+    }
+
+    onGroupsChange(newGroups);
+  },
+
+  handleRemoveGroupClick(groupIdx: number) {
+    if (!groups[groupIdx]) {
+      return;
+    }
+
+    const remainingGroups = [...groups];
+    remainingGroups.splice(groupIdx, 1);
+    onGroupsChange(remainingGroups);
+  },
+
+  handleAddGroupClick() {
+    if (groups.length > 1) {
+      // There can only be two groups ($and and $or)
+      // at any nesting level.
+      return;
+    }
+
+    const existingGroup = groups[0];
+    const alreadySelectedOperator = existingGroup?.logicalOperator;
+    const newGroupOperator =
+      alreadySelectedOperator === '$and' ? '$or' : '$and';
+    onGroupsChange([
+      ...groups,
+      createGroup({ logicalOperator: newGroupOperator }),
+    ]);
+  },
+});
+
+// Components - GroupHeader
+const groupHeaderStyles = css({
+  display: 'flex',
+  justifyContent: 'flex-start',
+  gap: spacing[3],
+  width: `calc(100% - ${CONDITION_CONTROLS_WIDTH})`,
+});
+
+export const GroupHeader = ({
+  dataTestId,
```

Comment:
We use `data-testid` elsewhere, I'd prefer we continue doing this so that they are easy to locate

Reply:
Addressed this as well in my recent commits.